UPGRADING.md

```diff
@@ -0,0 +1,516 @@
+    }),
+  },
+  {
+    path: '/contact-us',
+    exact: true,
+    component: asyncComponent({
+      loader: () => import('./Contact'),
+    }),
+  },
+];
+```
+
+To:
+
+```jsx
+// routes.js
+
+import Home from './Home';
+import { asyncComponent } from '@jaredpalmer/after';
+
+export default [
+  {
+    path: '/',
+    exact: true,
+    component: Home,
+  },
+  {
+    path: '/about',
+    exact: true,
+    component: asyncComponent({
+      loader: () => import(/* webpackChunkName: "whatever" */ './About'),
+      chunkName: 'whatever',
+    }),
+  },
+  {
+    path: '/contact-us',
+    exact: true,
+    component: asyncComponent({
+      loader: () => import(/* webpackChunkName: "ContactUs" */ './Contact'),
+      chunkName: 'ContactUs',
+    }),
+  },
+];
+```
+
+> This is too hard and complicated (and there are more things that we have to take care about) so we made a babel plugin to do this automatically because we care about you (DX).
+
+#### Use babel plugin
+
+Create a `.babelrc` file in the root of the project (next to the package.json)
+
+```js
+// .babelrc
+
+{
+  "presets": [ "razzle/babel" ],
+  "plugins": [ "after-async-component" ]
+}
+```
+
+> Check [after-async-component](https://github.com/nimacsoft/babel-plugin-after-async-component) repo for more options.
+
+**by enabling plugin THERE IS NO NEED TO CHANGE your routes, and old `route.js` file works.**
+
+#### Limitations of babel plugin
+
+There are some limitation, for example blow code won't work with plugin:
+
+```jsx
+import Home from './Home';
+import { asyncComponent } from '@jaredpalmer/after';
+
+const routes = [
+  {
+    path: '/',
+    exact: true,
+    component: Home,
+  },
+  {
+    path: '/about',
+    exact: true,
+    name: "About",
+    }),
+  },
+  {
+    path: '/contact-us',
+    exact: true,
+    name: "Contact",
+  }
+];
+
+function myTransformations(route) {
+  if (!route.name) return route
+  return {
+    ...route,
+    component: asyncComponent({
+      loader: () => import(`./pages/${route.name}`)
+    }),
+  }
+}
+
+export default routes.map(myTransformations)
+```
+
+to fix this just change `myTransformations` function implementaion to:
+
+```javascript
+function myTransformations(route) {
+  if (!route.name) return route;
+  return {
+    ...route,
+    component: asyncComponent({
+      loader: () =>
+        import(
+          /* webpackChunkName: "[request]" */
+          `./pages/${route.name}`
+        ),
+      chunkName: route.name,
+    }),
+  };
+}
+```
+
+### Add new webpack plugin
+
+We have to add new webpack plugin to get chunks information from webpack (it's called stats) and save it into a file to use it later in our application.
+
+This is not magic but it's using webpack magic comments.
+
+This will create a file called `manifest.json` in the build directory.
+
+to enable `razzle-plugin-manifest` create a `razzle.config.js` file in the root directory of the project (next to the package.json):
```

Comment:
Why not just contribute this as a PR to Razzle?

Reply:
yes, for sure.
should I remove `AssetsPlugin` from razzle webpack config, or should I add this alongside with that plugin?

if we remove `AssetsPlugin` there are going to be some breaking changes.

`AssetsPlugin` is going to generate this:
```json
{
  "client": {
    "css": "/static/css/bundle.36d04d42.css",
    "js": "/static/js/bundle.d3433edb.js"
  },
  "home": {
    "css": "/static/css/home.2f97ec2d.chunk.css",
    "js": "/static/js/home.82beecb0.chunk.js"
  },
  "about": {
    "js": "/static/js/about.1e639ac9.chunk.js"
  },
  "": {
    "js": ["/./static/js/sw.js", "/static/js/precache-manifest.96974c025fec5b5d12bd3959ac555f0b.js"],
    "png": ["/static/media/AviraLogo.906ef92a.png", "/static/media/BitdefenderLogo.57e13410.png", "/static/media/ESETLogo.fed44a25.png", "/static/media/FSecureLogo.9f9b31be.png", "/static/media/KasperskyLogo.6adab046.png", "/static/media/PadvishLogo.4ee6e526.png", "/static/media/PandaLogo.50c3b8b0.png", "/static/media/business.de4a6778.png", "/static/media/f-secure.62f6b90c.png", "/static/media/f-secure@2x.9aedf9f8.png", "/static/media/f-secure@3x.0b09d093.png", "/static/media/logo.299099e5.png", "/static/media/logo@2x.61bf8489.png", "/static/media/logo@3x.38c6193a.png", "/static/media/ransaomeware.69b525c3.png"],
    "woff2": ["/static/media/IRANSansWeb.0b5055ac.woff2", "/static/media/IRANSansWeb_Bold.43a0ecf3.woff2", "/static/media/IRANSansWeb_Medium.9c66b762.woff2"],
    "svg": ["/static/media/SliderButton.b3f3f52d.svg", "/static/media/add.587c8d1d.svg", "/static/media/addShoppingCart.839cb297.svg", "/static/media/arrowBottom.920f1bc1.svg", "/static/media/basket.14c8e15d.svg", "/static/media/basketFill.237a35f6.svg", "/static/media/bin.657d259e.svg", "/static/media/business.429de9e8.svg", "/static/media/clock.656a2e61.svg", "/static/media/delete.b91c4e05.svg", "/static/media/delivery-truck.ba934bc6.svg", "/static/media/discount.42ff889b.svg", "/static/media/edit.cc664895.svg", "/static/media/emptyCart.7acadfda.svg", "/static/media/fullStar.663eade5.svg", "/static/media/guarantee.6d0c5a7c.svg", "/static/media/instagram.6c1afdb6.svg", "/static/media/keyboardArrowLeft.dd3dbb8f.svg", "/static/media/power.6f864b09.svg", "/static/media/searchIcon.973823bd.svg", "/static/media/sub.86288a34.svg", "/static/media/twitter.a036d22e.svg"],
    "jpg": ["/static/media/bitdefender-off.c6291287.jpg", "/static/media/bitdefender.b4f9c3be.jpg", "/static/media/bitdefender@2x.818eb8a3.jpg", "/static/media/bitdefender@3x.35a4bbd8.jpg", "/static/media/blog-kaspersky.28417822.jpg", "/static/media/blog-padvish.629b86c4.jpg", "/static/media/kaspersky.2207b709.jpg", "/static/media/kaspersky@2x.2327672e.jpg", "/static/media/kaspersky@3x.26ef316b.jpg", "/static/media/userPlaceholder.02448a65.jpg"],
    "webp": ["/static/media/bitdefender.4ea5c386.webp", "/static/media/bitdefender@2x.44807f4c.webp", "/static/media/bitdefender@3x.44807f4c.webp", "/static/media/business.b0b5b473.webp", "/static/media/f-secure.7544497b.webp", "/static/media/f-secure@2x.bc7e6544.webp", "/static/media/f-secure@3x.765e0f60.webp", "/static/media/kaspersky.21d62015.webp", "/static/media/kaspersky@2x.386be870.webp", "/static/media/kaspersky@3x.c0c5e1a2.webp", "/static/media/logo.e662883a.webp", "/static/media/logo@2x.4e076e03.webp", "/static/media/logo@3x.8876d1f2.webp", "/static/media/ransaomeware.e3f9d29f.webp"]
  }
}
```
what this plugin will generate:
```json
{
  "client": {
    "css": [
      "/static/css/bundle.36d04d42.css"
    ],
    "js": [
      "/static/js/bundle.d3433edb.js",
      "/static/js/bundle.d3433edb.js.map"
    ]
  },
  "home": {
    "css": [
      "/static/css/home.2f97ec2d.chunk.css"
    ],
    "js": [
      "/static/js/home.82beecb0.chunk.js",
      "/static/js/home.82beecb0.chunk.js.map"
    ]
  },
  "about": {
    "css": [],
    "js": [
      "/static/js/about.1e639ac9.chunk.js",
      "/static/js/about.1e639ac9.chunk.js.map"
    ]
  }
}
```

the problem is not about that last object which has fonts, images, and other assets ...
my plugin saves `"client"` and other chunks CSS and JS file paths inside an array but `AssetsPlugin` saves them inside an object.

there is no way to handle them inside an object because of vendor chunks.

```js
"pages-Shop": {
  "css": [
    "/static/css/pages-Shop.0b5867d7.chunk.css"
  ],
  "js": [
    "/static/js/pages-Blog~pages-Shop.9dcd31d4.chunk.js", // <- shared js logic 
    "/static/js/pages-Shop.4409b1b8.chunk.js"
  ]
},
"pages-Blog": {
  "css": [
    "/static/css/pages-Blog.0c6823e6.chunk.css"
  ],
  "js": [
    "/static/js/pages-Blog~pages-Shop.9dcd31d4.chunk.js", // <- shared js logic
    "/static/js/pages-Blog.09994983.chunk.js"
  ]
```
I used a small trick to detect which vendor chunk is related to which chunk. inside the plugin, I saved the chunk name inside the vendor chunk file name and separated them with `~`.

Reply:
Yeah. I would keep both so we don't break things. AFAICT, there is no downside to having both. We will just want to update all the razzle examples with the new asset loop generated via the manifest plugin.

Reply:
I finally did it.
[https://github.com/jaredpalmer/razzle/pull/1178](https://github.com/jaredpalmer/razzle/pull/1178)